File: 13.Linked_List/python/8.DeletingNodeWithoutHead.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def deleteNode(node):
    # Copy next node's data to current node and bypass the next node
    if node and node.next:
        node.val = node.next.val
        node.next = node.next.next

def printLinkedList(head):
    current = head
    while current:
        print(current.val, end=" ")
        current = current.next
    print()

# Hardcoded linked list: 1 -> 2 -> 3
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)

# Link nodes
node1.next = node2
node2.next = node3

# Delete node with value 2 (simulate being given only the node2 reference)
deleteNode(node2)

# Print updated list: should be 1 -> 3
printLinkedList(node1)

File: 13.Linked_List/javascript/8.DeletingNodeWithoutHead.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

function buildLinkedList(arr) {
    if (arr.length === 0) return null;

    let head = new Node(arr[0]);
    let current = head;

    for (let i = 1; i < arr.length; i++) {
        current.next = new Node(arr[i]);
        current = current.next;
    }
    return head;
}

function deleteByValueWithoutHead(head, value) {
    let current = head;

    while (current !== null) {
        if (current.data === value && current.next !== null) {
            current.data = current.next.data;
            current.next = current.next.next;
            return;
        }
        current = current.next;
    }
    // If it's the tail or not found, we can't delete this way
}

function printList(head) {
    let current = head;
    let res = '';
    while (current !== null) {
        res += current.data + ' ';
        current = current.next;
    }
    console.log(res.trim());
}

let arr = [1, 2, 3]
let head = buildLinkedList(arr);
let valueToDelete = 2;
deleteByValueWithoutHead(head, valueToDelete);
printList(head);  // Output: 1 3
